### PR TITLE
Add custom port support to LocalPubsubHelper

### PR DESCRIPTION
**Pull request by shinfan**

### CODE REVIEW

File: gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/testing/LocalPubsubHelper.java [hunk 1]
```diff
@@ -35,63 +35,78 @@
  * A class that runs a Pubsub emulator instance for use in tests.
  */
 public class LocalPubsubHelper {
+
+  private final int port;
   private final LocalServiceHelper serviceHelper;
-  private final List<String> gcloudCommand;
-  private final URL emulatorUrl;
+
```
Comment (mziccard) [inline gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/testing/LocalPubsubHelper.java]: I think this empty line can be removed

File: gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/testing/LocalPubsubHelper.java [hunk 2]
```diff
@@ -35,63 +35,78 @@
...
 
   // Local server settings
   private static final int DEFAULT_PORT = 8080;
   private static final String DEFAULT_HOST = "localhost";
+  private static final URL EMULATE_URL;
```
Comment (mziccard) [inline gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/testing/LocalPubsubHelper.java]: Nit, I slightly prefer `EMULATOR_URL` (which is also in line with the previous name).

File: gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/testing/LocalPubsubHelper.java [hunk 3]
```diff
@@ -104,8 +119,8 @@ public void stop() throws IOException, InterruptedException {
    * Creates a channel for making requests to the in-memory service.
    */
   public ManagedChannel createChannel() {
-    return NettyChannelBuilder.forAddress(DEFAULT_HOST, DEFAULT_PORT)
+    return NettyChannelBuilder.forAddress(DEFAULT_HOST, port)
         .negotiationType(NegotiationType.PLAINTEXT)
         .build();
   }
-}
+}
```
Comment (mziccard) [inline gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/testing/LocalPubsubHelper.java]: I think your IDE removed the newline, could you please add it back?

**Comment by shinfan** [comment]
PTAL

File: gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/testing/LocalPubsubHelper.java [hunk 4]
```diff
@@ -35,63 +35,78 @@
...
+    port = LocalServiceHelper.findAvailablePort(DEFAULT_PORT);
+    List<String> gcloudCommand = new ArrayList<>(Arrays.asList(GCLOUD_CMD_TEXT.split(" ")));
+    gcloudCommand.add(GCLOUD_CMD_PORT_FLAG + port);
     GCloudEmulatorRunner gcloudRunner =
         new GCloudEmulatorRunner(gcloudCommand, VERSION_PREFIX, MIN_VERSION);
     DownloadableEmulatorRunner downloadRunner =
-        new DownloadableEmulatorRunner(Arrays.asList(BIN_NAME), emulatorUrl, MD5_CHECKSUM);
+        new DownloadableEmulatorRunner(Arrays.asList(BIN_NAME, BIN_CMD_PORT_FLAG + port),
+                                       EMULATE_URL,
+                                       MD5_CHECKSUM);
```
Comment (mziccard) [inline gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/testing/LocalPubsubHelper.java]: Nit: in `gcloud-java` we prefer a different wrapping usually:
```java
DownloadableEmulatorRunner downloadRunner =
    new DownloadableEmulatorRunner(Arrays.asList(BIN_NAME, BIN_CMD_PORT_FLAG + port),
        EMULATE_URL, MD5_CHECKSUM);
```
But if you used this also elsewhere feel free to keep it.

**Comment by mziccard** [comment]
Added a bunch of nits.

Also please @shinfan could you rebase these changes so that we can avoid the merge commit in this PR (1c9065e)?

**Comment by shinfan** [comment]
Also please @shinfan could you rebase these changes so that we can avoid the merge commit in this PR (1c9065e)?

I am planning to do a squash & merge. Does it make any difference if I rebase in this case? All sub commits will be squashed into one single PR commit.

**Comment by mziccard** [comment]
> I am planning to do a squash & merge. Does it make any difference if I rebase in this case? All sub commits will be squashed into one single PR commit.

No rebase is needed then. Just please avoid merges in PRs if not strictly necessary as it makes harder to understand what's going on. Looks good to me, feel free to merge. Thanks for doing that!

File: gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/testing/LocalPubsubHelper.java [hunk 5]
```diff
@@ -35,63 +35,76 @@
...
   // GCloud emulator settings
-  private static final String GCLOUD_CMD_TEXT = "gcloud beta emulators pubsub start --host-port";
+  private static final String GCLOUD_CMD_TEXT = "gcloud beta emulators pubsub start";
+  private static final String GCLOUD_CMD_PORT_FLAG = "--host-port=";
   private static final String VERSION_PREFIX = "pubsub-emulator";
   private static final String MIN_VERSION = "2016.01.13";
+  private static final String BIN_CMD_PORT_FLAG = "--port=";
 
   // Downloadable emulator settings
   private static final String FILENAME = "pubsub-emulator-20160113-2.zip";
```
Comment (aozarov) [inline gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/testing/LocalPubsubHelper.java]: BTW, There is a newer version: `tools/pubsub-emulator-20160113-3.zip`

**Pull request merged**